src/dotty/tools/dotc/transform/CheckReentrant.scala

```diff
@@ -0,0 +1,99 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import Names._
+import dotty.tools.dotc.transform.TreeTransforms.{AnnotationTransformer, TransformerInfo, MiniPhaseTransform, TreeTransformer}
+import ast.Trees._
+import Flags._
+import Types._
+import Constants.Constant
+import Contexts.Context
+import Symbols._
+import SymDenotations._
+import Decorators._
+import dotty.tools.dotc.core.Annotations.ConcreteAnnotation
+import dotty.tools.dotc.core.Denotations.SingleDenotation
+import scala.collection.mutable
+import DenotTransformers._
+import typer.Checking
+import Names.Name
+import NameOps._
+import StdNames._
+
+
+/** A no-op transform that checks whether the compiled sources are re-entrant.
+ *  If -Ycheck:reentrant is set, the phase makes sure that there are no variables
+ *  that are accessible from a global object. It excludes from checking paths that
+ *  are labeled with one of the annotations
+ *
+ *      @sharable   Indicating a class or val can be safely shared
+ *      @unshared   Indicating an object will not be accessed from multiple threads
+ *
+ *  Currently the analysis is only intended to check the dotty compiler itself. To make
+ *  it generally useful we'd need to add at least the following:
+ *
+ *   - Handle polymorphic instantiation: We might instantiate a generic class
+ *     with a type that contains vars. If the class contains fields of the generic
+ *     type, this may constitute a path to a shared var, which currently goes undetected.
+ *   - Handle arrays: Array elements are currently ignored because they are often used
+ *     in an immutable way anyway. To do better, it would be helpful to have a type
+ *     for immutable array.
+ */
+class CheckReentrant extends MiniPhaseTransform { thisTransformer =>
+  import ast.tpd._
+
+  override def phaseName = "checkReentrant"
+
+  def transformInfo(tp: Type, sym: Symbol)(implicit ctx: Context): Type = tp
```

Comment:
`ClassOf` does not implement `InfoTransformer` either and yet it seems to work, I don't know why: https://github.com/lampepfl/dotty/blob/6a36dd83e78fe2d6269626480577f8166ab7e37e/src/dotty/tools/dotc/transform/ClassOf.scala#L19